### PR TITLE
feat(sdk): model VL encoder data parallelism

### PR DESCRIPTION
**Pull request by wenpengw-nv**
#### Overview:

Model vision-encoder **data parallelism** for VL models and make it the default, mirroring vLLM (`--mm-encoder-tp-mode data`) and SGLang (`--mm-enable-dp-encoder`): ull ViT replica per rank, images ceil-split across the tp_size ranks, one exit AllGather instead of 2-per-layer AllReduce.
#### Details:
  - `fix(sdk)`: model the eager ViT RoPE-apply as a standalone encoder op.
  - `feat(sdk)`: `ModelConfig.enable_encoder_dp` (default True) — full-replica ops
    + exit `encoder_dp_all_gather`; the busiest rank (`ceil(images/tp)`) gates the
    phase; memory accounts replica weights + per-rank activations.
    `False` keeps identical op structure; absolute numbers shift due to the RoPE and memory fixes.
  - `feat(cli)`: `--disable-encoder-dp` on `default`/`estimate`; summary box shows
    `Encoder parallel: DP/TP`.

 Note: generator deploy configs don't emit the engine flags yet (follow-up); add `--mm-encoder-tp-mode data` / `--mm-enable-dp-encoder` manually to match the estimate.

#### Where should the reviewer start?

  - `aic-core/.../sdk/models/vit_ops.py` — op construction, AllGather payload
  - `aic-core/.../sdk/backends/base_backend.py` — `_run_encoder_phase` image split + encoder memory
#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- closes GitHub issue: #xxx

## Summary by CodeRabbit

* **New Features**
  * Added vision encoder data-parallel modeling by default (with an option to model legacy TP-style sharding via the new `--disable-encoder-dp` / `enable_encoder_dp` setting).
  * Added support for an optional external RoPE apply step in the vision encoder pipeline.
* **Bug Fixes**
  * Improved encoder latency/memory estimation for sharded image processing, including projected-embedding activations.
* **Documentation**
  * Clarified encoder DP behavior and how projector dimensions are interpreted.
* **Tests**
  * Added and updated unit tests to verify `--disable-encoder-dp` forwarding and propagation.

### CODE REVIEW

**Review by Arsene12358 — CHANGES_REQUESTED** [review]
Reviewed at 86303f34. First: the modeling core checks out — I verified it empirically on an editable install with the `h200_sxm` perf data (ceil image-sharding, tp=1 replica ops with the per-layer AllReduce no-op short-circuit, AllGather payload following the repo's total-gathered-volume convention from `context_cp_all_gather`, and the projected-embeddings memory term, whose measured delta of 0.625 GiB matches `2 B × 16384 tokens × 5120 × 4 instances` exactly). DP comes out ~27% faster than TP as expected, and text-only estimates are byte-identical to main. Nice work on the modeling itself. Requesting changes on two integration issues around it.

## B1 (blocking) — AFD static complement silently drops `enable_encoder_dp`

**Evidence chain**

1. `cli_estimate(mode="afd", afd_phase="decode", afd_combined_with_pd=True)` models the prefill side through `_run_static_estimate(static_mode="static_ctx", …)`, and that call forwards `image_height` / `image_width` / `num_images` — so the encoder IS modeled — but not `enable_encoder_dp`:

https://github.com/ai-dynamo/aiconfigurator/blob/86303f34a538a27a01cba290a8a57e2b06ca75ac/src/aiconfigurator/cli/api.py#L1058-L1070

2. `_run_static_estimate` declares `enable_encoder_dp=True`, so the omission silently re-enables DP modeling even when the user passed `--disable-encoder-dp`:

https://github.com/ai-dynamo/aiconfigurator/blob/86303f34a538a27a01cba290a8a57e2b06ca75ac/src/aiconfigurator/cli/api.py#L1244-L1248

3. Net effect: for a VL model in AFD decode mode, the prefill complement ignores the flag while the `agg` / `disagg` / `static*` paths honor it (api.py L870 / L904 / L957) — one run mixes both encoder-parallelism assumptions. (`afd_phase="prefill"` is unaffected: its complement is `static_gen`, which skips the encoder phase.)

**Proposed fix** — one line:

```diff
             num_images=num_images,
+            enable_encoder_dp=enable_encoder_dp,
             batch_size=batch_size,
```

Hardening (recommended): drop the `=True` defaults from the internal keyword-only helpers `_run_agg_estimate` / `_run_static_estimate` / `_run_disagg_estimate` (every real caller passes the value) and from `build_encoder_ops` (both Qwen3VL call sites pass it explicitly). With the parameter required, this omission would have surfaced as a `TypeError` in the AFD e2e path instead of a silent wrong default, and future VL model classes can't silently inherit DP either. A small regression test pinning AFD + `--disable-encoder-dp` (assert the complement's encoder numbers match the TP path) would close the loop.

## B2 (blocking) — DP-by-default estimates diverge from what generated deploy configs will actually run

**Evidence chain**

1. Measured on this branch (`h200_sxm`, trtllm 1.3.0rc10, tp=4, bs=8, 2×1024×1024 images, static mode):

```
aiconfigurator cli estimate --model-path Qwen/Qwen3-VL-32B-Instruct --system h200_sxm \
    --isl 4000 --osl 1000 --image-height 1024 --image-width 1024 --num-images 2 \
    --tp-size 4 --batch-size 8 --estimate-mode static [--disable-encoder-dp]
```

| | default (DP) | `--disable-encoder-dp` (TP) |
|---|---|---|
| Encoder latency | 99.451 ms | 136.856 ms |
| TTFT | 1576.274 ms | 1613.678 ms |
| Encoder memory | 1.830 GB | 1.322 GB |

2. Both engines default to the TP-sharded encoder: vLLM's `mm_encoder_tp_mode` defaults to `"weights"` ([vLLM optimization docs](https://docs.vllm.ai/en/stable/configuration/optimization/)) and SGLang's `--mm-enable-dp-encoder` is opt-in ([SGLang docs](https://docs.sglang.io/advanced_features/dp_for_multi_modal_encoder.html)).
3. The generator emits neither flag: `grep -rn "mm-encoder-tp-mode\|mm-enable-dp-encoder" src/aiconfigurator/generator/` → zero hits at 86303f34 (the only mentions in the tree are CLI help text).
4. Net: an out-of-the-box VL run prints `Encoder parallel: DP` and a ~99 ms encoder, but deploying the generated config on vLLM/SGLang yields the TP behavior this PR's own model puts at ~137 ms — TTFT ~2.3% optimistic at this workload, growing with image count/resolution. The description's "add the flags manually" note covers CLI users who read it, but defaults are exactly what people don't read release notes for, and webapp/generator consumers never see the CLI note at all.
5. trtllm (the default backend): the [TRT-LLM multimodal README](https://github.com/NVIDIA/TensorRT-LLM/blob/main/examples/models/core/multimodal/README.md) describes the visual encoder as replicated per GPU running data-parallel, so default-DP may already match there — but could you confirm whether the PyTorch flow actually shards images across TP ranks (true DP) or replicates full-batch ViT compute on every rank? If the latter, even the DP estimate is ~tp× optimistic on encoder compute for trtllm and the default should be per-backend.

**Proposed fix — either of these unblocks:**

**Option A (preferred): emit the engine flags from the generator in this PR.** The existing plumbing makes it mechanical rather than a heavy lift: (1) `generator/module_bridge.py::_build_worker_params` — add `enable_encoder_dp` from `task_config` to the worker payload for VL tasks (the `getattr(task_config, "image_height", 0)` VL-gate pattern already exists in this file, and `task_v2.Task` already carries the field after this PR); (2) `generator/config/backend_config_mapping.yaml` — map the param key per backend (`vllm: mm-encoder-tp-mode` with value `data`, `sglang: mm-enable-dp-encoder` as a boolean); (3) append the flag in the supported per-version `cli_args.<version>.j2` templates following the existing `enable-expert-parallel` boolean pattern — the per-version templates give engine-version gating for free. Estimate and deployment then agree by construction, and the DP default becomes the right default.

**Option B: keep the feature opt-in until the generator follow-up lands.** Flip the default to `False` at the default sites (`ModelConfig`, `config_builders.build_model_config`, `vit_ops.build_encoder_ops`, `task_v2.Task`, `cli_default` / `cli_estimate`) and make the CLI flag opt-in (`--enable-encoder-dp`), then flip both back together with generator emission. Costs nothing — the whole DP machinery stays in place behind the flag — and estimate↔deployment fidelity is preserved in the meantime.

---

**Non-blocking footnote**, so nobody trips over the numbers above: with `--disable-encoder-dp` the results are close to but not identical to pre-PR main (encoder 136.856 ms vs 134.263 ms; encoder memory 1.322 GB vs 0.697 GB) — that's the bundled RoPE-apply fix plus the new embeddings-buffer memory term, both intentional. Suggest rewording the description's "`False` keeps the legacy TP path bit-identical" to "identical op structure; absolute numbers shift due to the RoPE and memory fixes", to keep future bisects honest.

🤖 Reviewed with [Claude Code](https://claude.com/claude-code)

**Comment by wenpengw-nv** [comment]
@Arsene12358 
**B1**: fixed, now a required kwarg on the internal `_run_*_estimate` helpers so a missed forward raises instead of defaulting.
**B2**: Taking Option A: Templates: vLLM 0.12.0–0.20.1; SGLang only 0.5.10.post1/0.5.11 (the flag first shipped in SGLang 0.5.10 — older versions reject it at startup). Documented in the CLI user guide (default + estimate).

**Comment by Arsene12358** [comment]
@wenpengw-nv could you take a look at the remaining comments from coderabbit and resolve them?

File: src/aiconfigurator/cli/main.py [hunk 1]
```diff
@@ -270,6 +271,12 @@ def _add_default_mode_arguments(parser):
     parser.add_argument(
         "--num-images", type=int, default=1, help="Number of images per request for vision-language models. Default: 1."
     )
+    parser.add_argument(
+        "--disable-encoder-dp",
+        action="store_true",
+        help="Model the vision encoder as TP-sharded instead of the default data-parallel "
+        "(vLLM mm_encoder_tp_mode='data' / SGLang --mm-enable-dp-encoder semantics).",
+    )
```
[the remark this answers is not in the document]
Comment (wenpengw-nv) [reply src/aiconfigurator/cli/main.py:329]: Already implemented in 56860c92: generated configs now emit vLLM `--mm-encoder-tp-mode data|weights` and SGLang `--mm-enable-dp-encoder` (>= 0.5.10 templates only; TRT-LLM has no encoder knob), so deployment matches the estimator.

File: aic-core/src/aiconfigurator_core/sdk/models/vit_ops.py [hunk 2]
```diff
@@ -122,6 +131,13 @@ def _vit_transformer_ops(enc_cfg: common.VisionEncoderConfig, tp_size: int) -> l
         ops.CustomAllReduce("encoder_ar_2", depth, h_vit, tp_size),
     ]
 
+    if enc_cfg.partial_rotary_factor > 0:
+        # the attention op's internal RoPE term is disabled in exchange (partial_rotary_factor=0.0).
+        rope_dim = 6 * (n_vit // tp_size) * head_size_vit
+        result.append(ops.ElementWise("encoder_rope_apply", depth, rope_dim, rope_dim, 0.8))
```
[the remark this answers is not in the document]
Comment (wenpengw-nv) [reply aic-core/src/aiconfigurator_core/sdk/models/vit_ops.py:139]: The factor is the engine's rotary-table parameter, not a rotated fraction: in both vLLM (`ApplyRotaryEmb`, cos contract = head_size//2, no pass-through) and SGLang (`cos = torch.cat([cos, cos])` + full-dim `rotate_half`), the kernel rotates all of Q/K. Scaling `rope_dim` by it would undercount the op 2x.
Clarified the field docstring in 913b1718.

**Comment by wenpengw-nv** [comment]
@Arsene12358 Done — both resolved.

**Review by Arsene12358 — APPROVED** [review]
